Feature: Add customization options notice

Mayisha opened the pull request:
Fixes #1719

### Description
Added a notice component for showing the customization options. Notice is dismissable and appears only when UPE is enabled in the main Stripe settings page.

Desktop view in `Payment Methods` tab inside the main Stripe settings page
![payment tab](https://user-images.githubusercontent.com/33387139/132300029-141519b3-c56e-4990-8277-2bb53f532a9e.png)

Desktop view in `Settings` tab inside the main Stripe settings page
![settings tab](https://user-images.githubusercontent.com/33387139/132300049-e7eb5648-a6ba-4573-b53b-4d1059123480.png)

Mobile view inside the main Stripe settings page
![mobile view customization options](https://user-images.githubusercontent.com/33387139/132300088-f31bdc09-fe1b-4321-8845-093d206b0a9f.png)

### Testing instructions

- Enable UPE checkout ([this](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/24277919aa7b9029c39ecd997dd133990e7f65c7/includes/class-wc-stripe-feature-flags.php#L24) function should return `true`) to make the notice visible
- Toggle the value of `wc_show_upe_customization_options_notice` in `wp_options` to manage the visibility of the notice from DB level.

### QA steps

- [ ] Notice is present when UPE checkout is enabled
- [ ] Notice is absent when UPE checkout is disabled
- [ ] Clicking the `close`/`cross` icon removes the notice
- [ ] Notice does not reappear on refresh once it has been dismissed

Mayisha commented:
I will add the tests after getting some feedback on the implementation.

frosso commented on client/settings/customization-option-notice/index.js:
```diff
@@ -0,0 +1,67 @@
+ */
+import UpeContext from '../upe-toggle/context';
+
+const NoticeWrapper = styled.div`
```
A `Notice` component already exists in `@wordpress/components`, it just needs some adjustments on the margins and the background color. Storybook: https://wordpress.github.io/gutenberg/?path=/story/components-notice--default - WCPay implementation: https://github.com/Automattic/woocommerce-payments/blob/develop/client/components/inline-notice/ . It also includes a `isDismissible` prop.

What do you think about reusing that one?

Mayisha replied:
Thanks for this. I did not know this exists but as it does reusing is the best approach.

frosso commented on client/settings/customization-option-notice/index.js:
```diff
@@ -0,0 +1,67 @@
+						'woocommerce-gateway-stripe'
+					),
+					components: {
+						customizationOptions: <b></b>,
```
(minor)
```suggestion
						customizationOptions: <b />,
```

or you could even make this more explicit for translators by renaming the `customizationOptions` tag to just `strong`

frosso commented on client/settings/customization-option-notice/index.js:
```diff
@@ -0,0 +1,67 @@
+	const {
+		dismissCustomizationOptionsNotice,
+		isUpeEnabled,
+		shouldShowCustomizationOptionsNotice,
```
The intent of the `UpeToggleContextProvider` is to _just_ toggle the UPE "preview" flag, so that we can more easily remove all the code later (when there's no longer going to be a need for a "preview".
Rather than creating a new controller and modifying the `UpeContext`, what do you think about using `useDispatch` to just update the option?

Something like this:
```
import React, { useContext, useCallback } from 'react';
import { useSelect, useDispatch } from '@wordpress/data';

...

const CUSTOMIZATION_OPTIONS_NOTICE_OPTION =
	'wcstripe_customization_options_notice_dismissed';

const CustomizationOptionNotice = () => {
	const { isUpeEnabled } = useContext( UpeContext );
	const isCustomizationOptionsNoticeVisible = useSelect( ( select ) => {
		const { getOption, hasFinishedResolution } = select(
			'wc/admin/options'
		);

		const hasFinishedResolving = hasFinishedResolution( 'getOption', [
			CUSTOMIZATION_OPTIONS_NOTICE_OPTION,
		] );
		const isOptionDismissed =
			getOption( CUSTOMIZATION_OPTIONS_NOTICE_OPTION ) === 'yes';

		return hasFinishedResolving && ! isOptionDismissed;
	} );

	const { updateOptions } = useDispatch( 'wc/admin/options' );

	const handleDismiss = useCallback( () => {
		updateOptions( {
			[ CUSTOMIZATION_OPTIONS_NOTICE_OPTION ]: 'yes',
		} );
	}, [ updateOptions ] );

	if ( ! isUpeEnabled || ! isCustomizationOptionsNoticeVisible ) {
		return null;
	}

    ...
};
```

This way there's no need to create a new controller (we can just use WP's internal functions by leveraging `@wordpress/data`), there's no need to modify the context provider (and its consumers), and the whole display/dismiss logic is colocated into this single component (without modifying `includes/admin/class-wc-stripe-settings-controller.php`)

frosso reviewed (commented):
Works well - a few comments in regards to the implementation

Mayisha committed:
Add 'CustomizationOptionNotice' component

Mayisha committed:
Add tests

Mayisha committed:
Add necessary mocks

frosso commented on client/settings/customization-option-notice/index.js:
```diff
@@ -0,0 +1,76 @@
+const NoticeWrapper = styled( Notice )`
+	background: #def1f7;
+	border-left: 4px solid #00aadc;
+	margin: 16px 0px;
```
When using `0`, the measurement unit becomes unnecessary

```suggestion
	margin: 16px 0;
```

frosso commented on client/settings/customization-option-notice/index.js:
```diff
@@ -0,0 +1,76 @@
+ * Internal dependencies
+ */
+import UpeToggleContext from '../upe-toggle/context';
+
+const NoticeWrapper = styled( Notice )`
+	background: #def1f7;
+	border-left: 4px solid #00aadc;
```
There's an `InlineNotice` component that already has some of these styles included - maybe even the padding for `&.is-dismissible {...}` could be moved there.

```suggestion
import InlineNotice from 'wcstripe/components/inline-notice';

const NoticeWrapper = styled( InlineNotice )`
```

frosso commented on client/settings/customization-option-notice/index.js:
```diff
@@ -0,0 +1,76 @@
+	background: #def1f7;
+	border-left: 4px solid #00aadc;
+	margin: 16px 0px;
+	padding-right: 12px;
+
+	&.is-dismissible {
+		padding-right: 12px;
+	}
```
The `padding-right` seems to be declared twice. Since the notice is dismissble, the first padding declaration doesn't have an effect.

```suggestion

	&.is-dismissible {
		padding-right: 12px;
	}
```

frosso commented on client/settings/customization-option-notice/index.js:
```diff
@@ -0,0 +1,76 @@
+		<NoticeWrapper isDismissible={ true } onRemove={ handleDismissNotice }>
+			{ interpolateComponents( {
+				mixedString: __(
+					'{{strong}}Where are customization options?{{/strong}} In the new checkout experience, payment method details are automatically displayed in your customers’ languages so you don’t have to worry about writing them manually.',
```
```suggestion
					'{{strong}}Where are the customization options?{{/strong}} In the new checkout experience, payment method details are automatically displayed in your customers’ languages so you don’t have to worry about writing them manually.',
```

frosso reviewed (commented):
I'm seeing some inconsistencies with the spacing when compared with the design.
It looks like the declared margin (`margin: 16px 0px;`) can be removed - the `InlineNotice` component already has `margin: 0`. This would simplify the styles to just:
```
const NoticeWrapper = styled( InlineNotice )`
	&.is-dismissible {
		padding-right: 12px;
	}
`;
```

(I also made some adjustments on the spacing between the payment method list and the UPE banner [here](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1801), so we'll have to see how they'll play together)

It's also possible to move the `&.is-dismissible {padding-right: 12px;}` styles to `client/components/inline-notice/style.scss`, simplifying it even further.

![Markup 2021-09-08 at 14 18 18](https://user-images.githubusercontent.com/273592/132572758-7215b579-2e36-4e8f-a230-5b88c09e3a1c.png)

Mayisha committed:
Update 'InlineNotice' component
- Fix class name
- Add style for dismissable notice

Mayisha committed:
Update 'CustomizationOptionNotice'
 - Fix notice text
 - Use 'InlineNotice' instead of 'Notice' component
 - Refactor styles

Mayisha commented:
Used the `InlineNotice` component with suggested styles changes.

The `payment tab` looks like following
![tab 1](https://user-images.githubusercontent.com/33387139/132666373-6cd02c5f-d4c1-4431-88e9-25315c1965fa.png)

The settings tab has bigger gaps on the top and bottom of the component but that is coming from the temporary card. It can be fixed when we implement the actual component in the placeholders
![tab 2](https://user-images.githubusercontent.com/33387139/132666394-f53cd581-d5dd-42ad-a488-425f77cd9c6a.png)

frosso reviewed (approved):
Thanks for the changes! Tested again, looks great! 👍
